Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/extractors/BaseSetExtractor.java

```diff
@@ -138,9 +138,9 @@ public BaseSetExtractor(ISetExtractorConfiguration config) {
         matches = RegExpUtility.getMatches(setWeekDayRegex, input);
         for (Match match : matches) {
             if (match != null) {
-                String trimedText = sb.delete(match.index, match.index + match.length).toString();
```

Comment:
This seemed to be causing failures in the course of french tests. I switched to not using the "mutable" string builder which is modified between different loop iterations.

Reply:
Can you mention which tests failed? It would be strange that other languages don't fail and now only French would.

Reply:
These are the tests which fail:

1. "Je vais partir 9am chaque lundis"
```
test[DateTimeRecognizer - French - SetExtractor - "Je vais partir 9am chaque lundis"](com.microsoft.recognizers.text.tests.datetime.DateTimeExtractorTest)  Time elapsed: 0.027 sec  <<< ERROR!
java.lang.StringIndexOutOfBoundsException
	at java.lang.AbstractStringBuilder.delete(AbstractStringBuilder.java:760)
	at java.lang.StringBuilder.delete(StringBuilder.java:244)
	at com.microsoft.recognizers.text.datetime.extractors.BaseSetExtractor.matchEach(BaseSetExtractor.java:141)
	at com.microsoft.recognizers.text.datetime.extractors.BaseSetExtractor.extract(BaseSetExtractor.java:42)
	at com.microsoft.recognizers.text.tests.datetime.DateTimeExtractorTest.extract(DateTimeExtractorTest.java:113)
	at com.microsoft.recognizers.text.tests.datetime.DateTimeExtractorTest.recognizeAndAssert(DateTimeExtractorTest.java:86)
	at com.microsoft.recognizers.text.tests.AbstractTest.test(AbstractTest.java:115)
```
2. "Je vais partir tous les matins"
```
test[DateTimeRecognizer - French - SetExtractor - "Je vais partir tous les matins"](com.microsoft.recognizers.text.tests.datetime.DateTimeExtractorTest)  Time elapsed: 0.028 sec  <<< ERROR!
java.lang.StringIndexOutOfBoundsException
	at java.lang.AbstractStringBuilder.delete(AbstractStringBuilder.java:760)
	at java.lang.StringBuilder.delete(StringBuilder.java:244)
	at com.microsoft.recognizers.text.datetime.extractors.BaseSetExtractor.matchEach(BaseSetExtractor.java:141)
	at com.microsoft.recognizers.text.datetime.extractors.BaseSetExtractor.extract(BaseSetExtractor.java:42)
	at com.microsoft.recognizers.text.tests.datetime.DateTimeExtractorTest.extract(DateTimeExtractorTest.java:113)
	at com.microsoft.recognizers.text.tests.datetime.DateTimeExtractorTest.recognizeAndAssert(DateTimeExtractorTest.java:86)
	at com.microsoft.recognizers.text.tests.AbstractTest.test(AbstractTest.java:115)
```

It could be an error in the French implementation, but I'm not convinced that the original implementation of this method matches what, say, its Python counterpart is doing.

```python
    for match in regex.finditer(self.config.set_each_regex, source):
            trimmed_source = source[0:match.start()] + source[match.end():]

            for extract_result in extractor.extract(trimmed_source, reference):
                if extract_result.start <= match.start() < extract_result.start + extract_result.length:
                    yield Token(extract_result.start, extract_result.start + extract_result.length + len(match.group()))

        for match in regex.finditer(self.config.set_week_day_regex, source):
            trimmed_source = source[0:match.start()] \
                + RegExpUtility.get_group(match, Constants.WEEKDAY_GROUP_NAME) + source[match.end():]

            for extract_result in extractor.extract(trimmed_source, reference):
                if extract_result.start <= match.start() and\
                        RegExpUtility.get_group(match, Constants.WEEKDAY_GROUP_NAME) in extract_result.text:
                    length = extract_result.length + 1
                    prefix = RegExpUtility.get_group(match, Constants.PREFIX_GROUP_NAME)
                    if prefix:
                        length += len(prefix)

                    yield Token(extract_result.start, extract_result.start + length)
```
Note that each pass through the for loop, the Python version re-extracts the substring from the (unmutated) input. This is different than the original Java version, which reassigns `sb` to a new `StringBuilder` with the _trimmed_ text for the next iteration of the loop. This results in the deletion of a range of indices from the string builder which **does not** correspond to the text that was matched.